### PR TITLE
feat(serializer): support for getSupportedTypes (symfony 6.3)

### PR DESCRIPTION
**Pull request by mrossard**
| Q             | A
| ------------- | ---
| Branch?       | 3.1
| Tickets       | Closes https://github.com/api-platform/core/issues/5610

This is a quick draft for completing/revisiting the work initiated here : https://github.com/api-platform/core/pull/5611 , which seems abandoned right now.

### CODE REVIEW

**Comment by john-dufrene-dev** [comment]
@mrossard 

With this commit, all depreciations disappear ?

**Comment by mrossard** [comment]
> @mrossard
> 
> With this commit, all depreciations disappear ?

it seems ok on my app, haven't tested everything yet.

**Comment by mrossard** [comment]
Actually quite a lot of this is covered by existing tests, so i'll mark this as ready for review!

**Review by dunglas — CHANGES_REQUESTED** [review]
Thank you for working on this!

All `hasCacheableSupportsMethod()` must also be deprecated (as in Symfony).

**Comment by mrossard** [comment]
> Thank you for working on this!
> 
> All `hasCacheableSupportsMethod()` must also be deprecated (as in Symfony).

Shouldn't they be kept until APIP stops supporting older versions of SF? Using them instead of just copying their return values might be a mistake on my part though - i thought it could serve as a reminder that they are still useful for older SF versions.

File: src/Elasticsearch/Serializer/ItemNormalizer.php [hunk 1]
```diff
@@ -90,6 +90,11 @@
         return DocumentNormalizer::FORMAT !== $format && $this->decorated->supportsNormalization($data, $format);
     }
 
+    public function getSupportedTypes($format): array
+    {
+        return $this->decorated->getSupportedTypes($format);
```
[the remark this answers is not in the document]
Comment (mrossard) [reply src/Elasticsearch/Serializer/ItemNormalizer.php:95]: Yes, missed this one, thanks.

**Comment by dunglas** [comment]
> Shouldn't they be kept until APIP stops supporting older versions of SF?

It's what we'll do, but we should trigger a deprecation anyway to be able to remove these methods when we'll drop support for old Symfony versions, as they are also part of our public API. Old versions of Symfony will still work with newer versions of API Platform, but a deprecation will be triggered (as excepted, because these methods really are deprecated :D).

Also, to prevent PHPStan errors, you can use this trick: https://github.com/symfony/serializer/blob/f783a7ce6a048393f14f30e1094619f7729b1eb0/Debug/TraceableNormalizer.php#L44

**Comment by mrossard** [comment]
> > Shouldn't they be kept until APIP stops supporting older versions of SF?
> 
> It's what we'll do, but we should trigger a deprecation anyway to be able to remove these methods when we'll drop support for old Symfony versions, as they are also part of our public API. Old versions of Symfony will still work with newer versions of API Platform, but a deprecation will be triggered (as excepted, because these methods really are deprecated :D).

ok, i'll add deprecations for this then (and replace the calls I added right away while I'm at it). Would version 3.2 be the correct target for these?

> Also, to prevent PHPStan errors, you can use this trick: https://github.com/symfony/serializer/blob/f783a7ce6a048393f14f30e1094619f7729b1eb0/Debug/TraceableNormalizer.php#L44

👍

**Review by dunglas — CHANGES_REQUESTED** [review]
Thanks for the changes, we're almost there!

To fix the remaining deprecation notices, you need to add the `@group legacy` annotation on tests testing the `hasCacheableSupportsMethod()` method (btw, it would be nice to add unit tests for `getSupportedTypes()`).

Normalizers from the TestBundle must also be upgraded to `getSupportedTypes()`.

**Comment by mrossard** [comment]
This version should be better. I didn't add tests for getSupportedTypes() (yet?), but as they just return static arrays and are by design covered by functionnal tests I'm not sure how important those would be....?

**Comment by melroy89** [comment]
Thank you so much for this PR!

**Comment by dunglas** [comment]
Thank you @mrossard and @Oipnet!

**Pull request merged**